actors/builtin/payment_channel/payment_channel_actor.go

```diff
@@ -0,0 +1,247 @@
+
+		ls, ok := st.LaneStates[sv.Lane]
+		// create voucher lane if it does not already exist
+		if !ok {
+			ls = new(LaneState)
+			ls.Redeemed = big.NewInt(0)
+			st.LaneStates[sv.Lane] = ls
+		}
+		if ls.Closed {
+			rt.Abort(5, "cannot redeem a voucher on a closed lane")
+		}
+
+		if ls.Nonce > sv.Nonce {
+			rt.Abort(6, "voucher has an outdated nonce, cannot redeem")
+		}
+
+		// The next section actually calculates the payment amounts to update the payment channel state
+		// 1. (optional) sum already redeemed value of all merging lanes
+		redeemedFromOthers := big.Zero()
+		for _, merge := range sv.Merges {
+			if merge.Lane == sv.Lane {
+				rt.Abort(7, "voucher cannot merge lanes into its own lane")
+			}
+
+			otherls := st.LaneStates[merge.Lane]
+
+			if otherls.Nonce >= merge.Nonce {
+				rt.Abort(8, "merged lane in voucher has outdated nonce, cannot redeem")
+			}
+
+			redeemedFromOthers = big.Add(redeemedFromOthers, otherls.Redeemed)
+			otherls.Nonce = merge.Nonce
+		}
+
+		// 2. To prevent double counting, remove already redeemed amounts (from
+		// voucher or other lanes) from the voucher amount
+		ls.Nonce = sv.Nonce
+		balanceDelta := big.Sub(sv.Amount, big.Add(redeemedFromOthers, ls.Redeemed))
+		// 3. set new redeemed value for merged-into lane
+		ls.Redeemed = sv.Amount
+
+		newSendBalance := big.Add(st.ToSend, balanceDelta)
+
+		// 4. check operation validity
+		if newSendBalance.LessThan(big.Zero()) {
+			rt.Abort(9, "voucher would leave channel balance negative")
+		}
+		if newSendBalance.GreaterThan(rt.CurrentBalance()) {
+			rt.Abort(10, "not enough funds in channel to cover voucher")
+		}
+
+		// 5. add new redemption ToSend
+		st.ToSend = newSendBalance
+
+		// update channel closingAt and MinCloseAt if delayed by voucher
+		if sv.MinCloseHeight != 0 {
+			if st.ClosingAt != 0 && st.ClosingAt < sv.MinCloseHeight {
+				st.ClosingAt = sv.MinCloseHeight
+			}
+			if st.MinCloseHeight < sv.MinCloseHeight {
+				st.MinCloseHeight = sv.MinCloseHeight
+			}
+		}
+		return nil
+	})
+	return &vmr.EmptyReturn{}
+}
+
+func (pca *PaymentChannelActor) Close(rt vmr.Runtime) *vmr.EmptyReturn {
+
+	var st PaymentChannelActorState
+	rt.State().Transaction(&st, func() interface{} {
+
+		rt.ValidateImmediateCallerIs(st.From, st.To)
+
+		if st.ClosingAt != 0 {
+			rt.Abort(1, "channel already closing")
+		}
+
+		st.ClosingAt = rt.CurrEpoch() + indices.PaymentChannel_PaymentChannelClosingDelay()
+		if st.ClosingAt < st.MinCloseHeight {
+			st.ClosingAt = st.MinCloseHeight
+		}
+
+		return nil
+	})
+	return &vmr.EmptyReturn{}
+}
+
+func (pca *PaymentChannelActor) Collect(rt vmr.Runtime) *vmr.EmptyReturn {
+
+	var st PaymentChannelActorState
+	rt.State().Transaction(&st, func() interface{} {
+
+		if st.ClosingAt == 0 {
+			rt.Abort(1, "payment channel not closing or closed")
+		}
+
+		if rt.CurrEpoch() < st.ClosingAt {
+			rt.Abort(2, "payment channel not closed yet")
+		}
+
+		// send remaining balance to "From"
+
+		_, code := rt.Send(
```

Comment:
Illegal Send inside transaction

Reply:
Understood. Nonetheless I worry about atomicity here. Has this been worked out elsewhere yet?

Reply:
Not sure I understand the question. When updating all the other actors to use transactions, I think I avoided illegal side effects, but unsure until we write tests.

Reply:
As in what if the balance gets sent out, and then the transaction fails below and payment channel stays funded even though it's already paid out (in this case)

Reply:
I can no longer link this comment to the code so I've lost context.

Note that if a method aborts, all state changes are rolled back, desipite the name transaction. Does that answer it?

Reply:
yes and I guess this is the crux of the issue. You should be able to somehow have the transaction rollback on send failures as well in that both states should be in sync